### PR TITLE
Dynamic log level support

### PR DESCRIPTION
**Pull request by taburet**
EE counterpart: https://github.com/hazelcast/hazelcast-enterprise/pull/3923

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 10 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/jmx/InstanceMBean.java:27:8: Unused import - com.hazelcast.logging.impl.LoggingServiceImpl. [UnusedImports]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/jmx/LoggingServiceMBean.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/logging/impl/InternalLogger.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/logging/impl/InternalLoggerFactory.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/logging/impl/LoggingServiceImpl.java:63:28: Variable 'levelSet' explicitly initialized to 'null' (default value for its type). [ExplicitInitialization]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/logging/JdkLoggerLevelChangeTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/logging/Log4j2LoggerLevelChangeTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/logging/Log4jLoggerLevelChangeTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/logging/NoLoggerLevelChangeTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/logging/Slf4jLoggerLevelChangeTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
</pre>
</details>

File: hazelcast/src/test/java/com/hazelcast/logging/Slf4jLoggerLevelChangeTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    @Before
+    public void before() {
+        isolatedLoggingRule.setLoggingType(LOGGING_TYPE_SLF4J);
+        HazelcastInstance instance = createHazelcastInstance();
+        loggingService = (LoggingServiceImpl) instance.getLoggingService();
+    }
+
+    @Test(expected = HazelcastException.class)
+    public void test() {
+        // slf4j is not supported currently
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/logging/Slf4jLoggerLevelChangeTest.java:54]: slf4j does not support changing log level at runtime?

**Review by vbekiaris — COMMENTED** [review]
Overall looks good, I will wait for the TDD to approve :-). Since we are exposing log level via REST API, I think it makes sense to add log level read/set/reset operation in `cluster.sh` script.

File: hazelcast/src/test/java/com/hazelcast/logging/Slf4jLoggerLevelChangeTest.java (hunk 1 above)
Comment (taburet) [reply hazelcast/src/test/java/com/hazelcast/logging/Slf4jLoggerLevelChangeTest.java:54]: Unfortunately, there is no public or internal API for changing the level in slf4j, so for each logging backend supported by slf4j a separate specialized code should be written to extract the loggers from internal slf4j wrappers. The problem is that the wrappers are internal and specific for every logging backend, so they are not necessarily stable across versions of slf4j. In theory, we could support some set of popular logging backends for slf4j, but that would require inspecting at least recent versions of slf4j and wrappers to understand how stable the implementation of each wrapper is and how we could access it, regularly or through reflection. After discussion with Jiri and Matko we decided to postpone that.

**Review by kwart — COMMENTED** [review]
Mostly LGTM. The only comments are related to handling the REST endpoint.

**Comment by taburet** [comment]
TDD added in https://github.com/hazelcast/hazelcast/pull/18127/commits/c8c414dafe456bed7a569f7324c6fdb44ea7e67f

**Comment by taburet** [comment]
@vbekiaris Regarding `cluster.sh`, I don't see any support in it for commands targeting a specific member, are you thinking about another REST endpoint to manage the log level in a cluster-wide manner?

File: hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpPostCommandProcessor.java [hunk 2]
```diff
@@ -567,4 +571,15 @@ protected void handleSetLicense(HttpPostCommand cmd) throws Throwable {
         // NO-OP in OS
         prepareResponse(cmd, response(SUCCESS));
     }
+
+    private void handleLogLevel(HttpPostCommand command) throws UnsupportedEncodingException {
+        String[] params = decodeParamsAndAuthenticate(command, 3);
+        String rawLevel = params[2];
+        Level level = Level.parse(rawLevel);
```
[the remark this answers is not in the document]
Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpPostCommandProcessor.java]: fixed in https://github.com/hazelcast/hazelcast/pull/18127/commits/603ddac55a182a8d01f63fc8de79b9edbb6b7ba9

File: hazelcast/src/test/java/com/hazelcast/internal/ascii/RestLogLevelTest.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Test
+    public void testDisabledRest() {
+        // REST should be disabled by default
+        HazelcastInstance instance = factory.newHazelcastInstance(createConfig());
+        HTTPCommunicator communicator = new HTTPCommunicator(instance);
+
+        try {
+            communicator.getLogLevel();
```
[the remark this answers is not in the document]
Comment (taburet) [reply hazelcast/src/test/java/com/hazelcast/internal/ascii/RestLogLevelTest.java]: fixed in https://github.com/hazelcast/hazelcast/pull/18127/commits/0ec33ac2741d248437231165229eae6bb1ed1e75

**Comment by vbekiaris** [comment]
> @vbekiaris Regarding `cluster.sh`, I don't see any support in it for commands targeting a specific member, are you thinking about another REST endpoint to manage the log level in a cluster-wide manner?

I mentioned `cluster.sh` because I think this is our only current "admin" tool, but you are right that `--operation` documentation states "Executes cluster-wide operation", so the current implementation doesn't seem to fit.

File: hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpDeleteCommandProcessor.java [hunk 4]
```diff
@@ -85,4 +90,11 @@ private void handleQueue(HttpDeleteCommand command, String uri) {
     public void handleRejection(HttpDeleteCommand command) {
         handle(command);
     }
+
+    private void handleLogLevel(HttpDeleteCommand command) {
```
[the remark this answers is not in the document]
Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpDeleteCommandProcessor.java:94]: Currently, the REST "framework" supports auth only for POST and there are no commands in GET and DELETE families supporting auth. One can even delete entire IMaps without any auth, reseting log level back to its default value without an auth doesn't look critical in that light. Bringing auth support to other HTTP methods basically means a major refactoring in REST which is not a goal of this PR, TDD or PRD.

[the remark this answers is not in the document]
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpDeleteCommandProcessor.java:94]: So what about handling the `resetLevel()` as an additional `POST` operation with another URL path?

File: docs/design/logging/01-dynamic-log-level.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    /**
+     * Sets the levels of all the loggers known to this logger factory to
+     * the given level. If a certain logger was already preconfigured with a more
+     * verbose level than the given level, it will be kept at that more verbose
+     * level.
+     *
+     * @param level the level to set.
+     */
+    void setLevel(@Nonnull Level level);
```
Comment (vbekiaris) [inline docs/design/logging/01-dynamic-log-level.md:57]: with latest push this is actually `setLevel(@Nonnull String level)`.

File: hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpDeleteCommandProcessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpDeleteCommandProcessor.java]: fixed in https://github.com/hazelcast/hazelcast/pull/18127/commits/182a44f43e1515b041a63a2036196cf42a41ed06

**Review by vbekiaris — APPROVED** [review]

**Comment by taburet** [comment]
@kwart all the comments are addressed, could you please inspect the latest changes?

**Review by kwart — APPROVED** [review]
:+1:

**Pull request merged**